sklearn/ensemble/_forest.py

```diff
@@ -599,14 +614,12 @@ def _compute_oob_predictions(self, X, y):
             n_samples,
             self.max_samples,
         )
-        for estimator in self.estimators_:
-            unsampled_indices = _generate_unsampled_indices(
-                estimator.random_state,
-                n_samples,
-                n_samples_bootstrap,
-            )
 
-            y_pred = self._get_oob_predictions(estimator, X[unsampled_indices, :])
+        all_pred = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
+                delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
+            for e in self.estimators_)
+
+        for unsampled_indices, y_pred in all_pred:
```

Comment:
Why not include all the other output below as part of the parallelization? It seems inefficient to have to loop twice over `all_pred`?

Reply:
I line profiled the code and identified the bottleneck, then scoped the part of the algorithm that made sense to make parallel.

With 200,000 features (1,000 records, 50 estimators) on 16 cores the finalization loop has negligible contribution so I didn't want to make code concurrent that didn't need to be:

```
   619         2     688182.1 344091.1     99.8          all_pred = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
   620                                                           delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
   621         1          0.2      0.2      0.0              for e in self.estimators_)
   622                                           
   623        51         26.0      0.5      0.0          for unsampled_indices, y_pred in all_pred:
   624        50       1111.8     22.2      0.2              oob_pred[unsampled_indices, ...] += y_pred
   625        50        386.2      7.7      0.1              n_oob_pred[unsampled_indices, :] += 1
```

If I bump that up to 5,000 estimators it is even more dramatic:

```
   619         2   69161602.1    3e+07     99.9          all_pred = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
   620                                                           delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
   621         1          0.2      0.2      0.0              for e in self.estimators_)
   622                                           
   623      5001        803.0      0.2      0.0          for unsampled_indices, y_pred in all_pred:
   624      5000      34326.1      6.9      0.0              oob_pred[unsampled_indices, ...] += y_pred
   625      5000      12399.4      2.5      0.0              n_oob_pred[unsampled_indices, :] += 1
```

Reply:
An alternative way that's more in-line w/ the existing `predict_proba` parallelization is

1. The following private function:

```
def _accumulate_oob_prediction(predict, tree, X, n_samples, n_samples_bootstrap, out, out_count, random_state, lock):
    """
    This is a utility function for joblib's Parallel.

    It can't go locally in ForestClassifier or ForestRegressor, because joblib
    complains that it cannot pickle it when placed there.
    """
    unsampled_indices = _generate_unsampled_indices(
        random_state,
        n_samples,
        n_samples_bootstrap,
    )

    prediction = predict(tree, X[unsampled_indices, :])
    with lock:
        out[unsampled_indices, ...] += prediction
        out_count[unsampled_indices, :] += 1
```

2. The `_compute_oob_predictions` LOC is replaced with:

```
       n_samples_bootstrap = _get_n_samples_bootstrap(
            n_samples,
            self.max_samples,
        )
        lock = threading.Lock()
        Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
            delayed(_accumulate_oob_prediction)(self._get_oob_predictions, e, X, n_samples, n_samples_bootstrap,
                                                oob_pred, n_oob_pred, e.random_state, lock)
            for e in self.estimators_
        )
```

However, I think one of the maintainers will chime in.

Reply:
I didn't see a reason to pass in a data structure and use a lock to accumulate in it. If there's a compelling performance argument I'd be happy to. I did see that infrastructure and decided it wasn't quite what I wanted, but I don't really care too much as long as the performance improvement holds steady.

Reply:
Given that this is using multithreading with `require="sharedmem"`, I think the memory issue comes from having to store all the predictions all at once in `all_preds`.

I suspect that having each worker update `oob_pred` with a lock would resolve the memory issue in https://github.com/scikit-learn/scikit-learn/pull/28442#issuecomment-2334910902

Reply:
I think the lock contention is too slow to scale well for our use case. I committed the diff below the fold locally and repeated the two benchmarking incantations, and the results were not particularly exciting. The memory cost is gone, but so is the scaling with more concurrency.

<details>

```diff
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -557,14 +557,16 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             (accuracy score).
         """
 
-    def _get_oob_pred_parallel(self, estimator, n_samples, n_samples_bootstrap, X):
+    def _get_oob_pred_parallel(self, oob_pred, n_oob_pred, estimator, n_samples, n_samples_bootstrap, X, lock):
         unsampled_indices = _generate_unsampled_indices(
             estimator.random_state,
             n_samples,
             n_samples_bootstrap,
         )
         y_pred = self._get_oob_predictions(estimator, X[unsampled_indices, :])
-        return unsampled_indices, y_pred
+        with lock:
+            oob_pred[unsampled_indices, ...] += y_pred
+            n_oob_pred[unsampled_indices, :] += 1
 
     def _compute_oob_predictions(self, X, y):
         """Compute and set the OOB score.
@@ -607,17 +609,12 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             self.max_samples,
         )
 
-        all_pred = Parallel(
-            n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem"
-        )(
-            delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
+        lock = threading.Lock()
+        Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
+            delayed(self._get_oob_pred_parallel)(oob_pred, n_oob_pred, e, n_samples, n_samples_bootstrap, X, lock)
             for e in self.estimators_
         )
 
-        for unsampled_indices, y_pred in all_pred:
-            oob_pred[unsampled_indices, ...] += y_pred
-            n_oob_pred[unsampled_indices, :] += 1
-
         for k in range(n_outputs):
             if (n_oob_pred == 0).any():
                 warn(
```

</details>

First incantation:

```
| Change   | Before [0b0b90b7] <main>   | After [264d74de] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                |
|----------|----------------------------|-----------------------------------------|---------|----------------------------------------------------------------------|
| -        | 1.16±0.02s                 | 838±10ms                                |    0.72 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 8, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Second incantation at higher concurrency/larger problems:

```
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

I guess I could try thinking about other models of concurrency a bit. I think when I played with other approaches they had other annoying tradeoffs, but maybe I should double check..

Reply:
I dunno, the other backends like `loky` do indeed seem to have serialization issues and so on. We could try to split the array and recombine it at the end--is that worth exploring?

Reply:
Ah I see. Can you try joblib's new generator support? The API is almost the same as what you already have but with `Parallel(..., return_as="generator")`

Reference: https://joblib.readthedocs.io/en/latest/auto_examples/parallel_generator.html